### PR TITLE
Add function for waiting on multiple wait sets together

### LINKED ISSUE #308: Wait set groups
## Feature request
This is a request for an API on top of the current waitset api that considers groups of waitable entities.

#### Feature description

An API that covers groups of entities that can be waited on would be immediately useful for actions. A server and client consists of several waitable entities (servers, clients, maybe a timer). The API would allow a client library to add a group of waitable entities to the wait set, and then notify the action server when one of them becomes available.

#### Implementation considerations

This could be implemented in a package `rcl_grouped_waitset` on top of the existing `rcl` API

Structs and methods to be added.
```C
struct rcl_grouped_waitset
{
  rcl_waitset_group_t * groups;
  size_t num_groups;

  rcl_waitset_group_impl_t * impl;
}

struct rcl_waitset_group_t
{
  rcl_subscription_t * subscriptions;
  rcl_guard_condition_t * guard_conditions;
  rcl_timer_t * timers;
  rcl_client_t * clients;
  rcl_service_t * servers;

  size_t num_subscriptions;
  size_t num_guard_conditions;
  size_t num_timers;
  size_t num_clients;
  size_t num_servers;
}
rcl_waitset_group_get_zero_initialized();

// User won't call these, but rcl_grouped_waitset_add and rcl_grouped_waitset_fini might
// rcl_waitset_group_init(rcl_waitset_group_t *, num_subs, subs, ..., rcl_allocator_t alloc);
// rcl_waitset_group_fini(rcl_waitset_group_t * g);

rcl_grouped_waitset_get_zero_initialized();
rcl_grouped_waitset_init(rcl_grouped_waitset_t * gws, size_t num_groups, rcl_allocator_t alloc);
rcl_grouped_waitset_fini(rcl_grouped_waitset_t * gws);

// Add a group waitable entities to the wait set
// This returns a waitset group that can be used to tell which entities in the group woke the wait set
// The returned group isn't usable until `rcl_grouped_waitset_build()` is called. The returned group never needs to be cleaned up by the user.
// subs, gcs, timers, etc pointers are borrowed and must be valid until `rcl_grouped_waitset_build()` is called
rcl_grouped_waitset_add(rcl_grouped_waitset_t * gws, rcl_waitset_group_t * g, size_t num_subs, rcl_subscription_t * subs, size_t num_guard_conditions, , rcl_guard_condition_t * gcs, size_t num_timers, rcl_timer_t * timers, size_t num_clients, rcl_client_t * clients, size_t num_servers, rcl_server_t * servers);

// Called after all groups have been added so allocation of the wait set only happens once
rcl_grouped_waitset_build();

// Calls `rcl_wait(rcl_waitset_t ws,...)` internally using a waitset built in `_build()` and stored in the implementation of the group object.
rcl_grouped_wait(rcl_grouped_waitset_t * gws, int64_t timeout);
```

Example use
```C
rcl_waitset_group_t gws = rcl_grouped_waitset_get_zero_initialized();
rcl_grouped_waitset_init(&gws, num_groups, alloc);

// Client library builds a group with all of the users subs, clients, services, timers, etc
rcl_waitset_group_t cl_group = rcl_waitset_group_get_zero_initialized();
rcl_grouped_waitset_add(&gws, &cl_group, ...);

// For each action server client library asks it to add itself to the waitset group
{
  rcl_waitset_group group = rcl_waitset_group_get_zero_initialized();
  // This calls rcl_grouped_waitset_add() internally
  rcl_action_server_add_to_grouped_waitset(&as, &gws, &group);
}

rcl_grouped_wait(gws, ...)

// Entities are set to NULL within the added groups
for (size_t g = 0; g < gws.num_groups; ++g)
{
  const rcl_waitset_group_t * group = gws.groups[g];
  for (size_t s = 0; s < group->num_subscriptions; ++s)
  {
    // client library does right thing
    // First group could be user defined stuff,
    // Following groups could be for action servers, action clients, etc
  }
}

// When done finalizing the grouped waitset also cleans up the groups
// This works because rcl_grouped_waitset_build() set the pointers to the entities to be internal to an `rcl_waitset_t1
rcl_grouped_waitset_fini(gws);
```

connects to ros2/rcl#307

### PR DESCRIPTION
**Pull request by jacobperron**
Internally, the function collates all wait sets into a single wait set that is waited on.

Resolves #308

### CODE REVIEW

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5546/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2200/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4575/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5436/)

File: rcl/include/rcl/wait.h [hunk 1]
```diff
@@ -428,6 +427,119 @@ RCL_WARN_UNUSED
 rcl_ret_t
 rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout);
 
+/// Block until multiple wait sets are ready or until the timeout has been exceeded.
+/**
+ * This function has similar behavior as rcl_wait(), but acts on multiple wait sets.
```
Comment (hidmic) [inline rcl/include/rcl/wait.h]: @jacobperron nit: similar behavior _to_?

File: rcl/test/rcl/test_wait.cpp [hunk 2]
```diff
@@ -435,3 +435,28 @@ TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), guard_condition) {
   EXPECT_EQ(RCL_RET_OK, f.get());
   EXPECT_LE(std::abs(diff - trigger_diff.count()), TOLERANCE);
 }
+
+// Test rcl_wait_multiple with a positive finite timeout value (1ms)
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), finite_timeout_multiple) {
```
Comment (hidmic) [inline rcl/test/rcl/test_wait.cpp]: @jacobperron shouldn't we add a few more tests? Like, for negative timeout, for zero timeout, for invalid arguments, for invalid wait sets, with a dummy guard and a dummy timer in different wait sets to make sure things are being collated and then updated back properly, and so on. Not to repeat `rcl_wait()` tests, just to verify the new functionality is correct.

File: rcl/src/rcl/wait.c [hunk 3]
```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
+  size_t i;
+  size_t j;
+  rcl_wait_set_t * wait_set;
+  // Count number of entities in wait sets while checking for invalid wait sets
+  size_t number_of_subscriptions = 0u;
+  size_t number_of_guard_conditions = 0u;
+  size_t number_of_timers = 0u;
+  size_t number_of_clients = 0u;
+  size_t number_of_services = 0u;
+  for (i = 0u; i < num_wait_sets; ++i) {
```
Comment (hidmic) [inline rcl/src/rcl/wait.c]: @jacobperron it could be that no wait set was provided, should we have a different rcl error message for that case (right now it'll say that wait sets are empty)?

File: rcl/src/rcl/wait.c [hunk 4]
```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
+    number_of_services += wait_set->size_of_services;
+  }
+  if (
+    number_of_subscriptions == 0u &&
+    number_of_guard_conditions == 0u &&
+    number_of_timers == 0u &&
+    number_of_clients == 0u &&
+    number_of_services == 0u)
+  {
+    RCL_SET_ERROR_MSG("wait set is empty");
```
Comment (hidmic) [inline rcl/src/rcl/wait.c]: @jacobperron consider `"wait sets are empty"` instead.

File: rcl/src/rcl/wait.c [hunk 5]
```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
   return RCL_RET_OK;
 }
 
+rcl_ret_t
+rcl_wait_multiple(rcl_wait_set_t wait_sets[], const size_t num_wait_sets, int64_t timeout)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_sets, RCL_RET_INVALID_ARGUMENT);
+  size_t i;
+  size_t j;
```
Comment (hidmic) [inline rcl/src/rcl/wait.c]: @jacobperron nit: I have a slight preference for declaring loop variables within the `for` statement. Won't block if you prefer to keep things C89-ish.

File: rcl/src/rcl/wait.c [hunk 6]
```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
+  // Collate all wait sets into a single wait set
+  for (i = 0u; i < num_wait_sets; ++i) {
+    wait_set = &wait_sets[i];
+    for (j = 0u; j < wait_set->size_of_timers; ++j) {
+      const size_t current_index = collated_wait_set.impl->timer_index++;
+      collated_wait_set.timers[current_index] = wait_set->timers[j];
+      rcl_guard_condition_t * guard_condition = rcl_timer_get_guard_condition(wait_set->timers[j]);
+      if (NULL != guard_condition) {
+        // rcl_wait() will take care of moving these backwards and setting guard_condition_count.
+        const size_t index = wait_set->size_of_guard_conditions + (wait_set->impl->timer_index - 1);
```
Comment (hidmic) [inline rcl/src/rcl/wait.c]: @jacobperron not sure I follow how `index` would take different values across multiple iterations for the same `wait_set`.

File: rcl/src/rcl/wait.c [hunk 7]
```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
+
+  // Update individual wait sets
+  size_t collated_subscription_index = 0u;
+  size_t collated_guard_condition_index = 0u;
+  size_t collated_timer_index = 0u;
+  size_t collated_client_index = 0u;
+  size_t collated_service_index = 0u;
+  for (i = 0u; i < num_wait_sets; ++i) {
+    wait_set = &wait_sets[i];
+    for (j = 0u; j < wait_set->size_of_timers; ++j) {
```
Comment (hidmic) [inline rcl/src/rcl/wait.c]: @jacobperron nit: fine to use `for` loops, but we could use `memcpy` too (that depending on the compiler and target architecture may be more performant).

File: rcl/src/rcl/wait.c (hunk 3 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: My intention was to return `RCL_RET_WAIT_SET_EMPTY` thinking it is intuitive, but I can add a new return code e.g. `RCL_RET_WAIT_SET_NONE`.

Comment (hidmic) [reply rcl/src/rcl/wait.c]: Either that or `RCL_RET_INVALID_ARGUMENT` plus `RCL_SET_ERROR_MSG()` with a nice message.

File: rcl/src/rcl/wait.c (hunk 5 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: I guess my head was in C89 mode. Since we're targeting C99, I'm happy to change it.

File: rcl/src/rcl/wait.c (hunk 7 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: Sure, I can do that. Do you think there is any issues with readability (and correctness)?

```c
    memcpy(
      wait_set->timers,
      collated_wait_set.timers + collated_timer_index,
      sizeof(rcl_timer_t *) * wait_set->size_of_timers);
```

Comment (hidmic) [reply rcl/src/rcl/wait.c]: Ugh, that doesn't look nearly as clean as your current code. Nevermind, let's trust loop unrolling will work :).

Comment (jacobperron) [reply rcl/src/rcl/wait.c]: :+1:

File: rcl/src/rcl/wait.c (hunk 6 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: It won't, this is a copy-paste bug.

File: rcl/test/rcl/test_wait.cpp (hunk 2 above)
Comment (jacobperron) [reply rcl/test/rcl/test_wait.cpp]: Yes, I meant to, but forgot.

**Comment by jacobperron** [comment]
@hidmic Thanks for the review! I think I've address all of your comments. Ready for another round of review.

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5561/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2207/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4582/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5450/)

File: rcl/src/rcl/wait.c [hunk 8]
```diff
@@ -614,6 +610,157 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
 }
 
+rcl_ret_t
+rcl_wait_multiple(rcl_wait_set_t wait_sets[], const size_t num_wait_sets, int64_t timeout)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_sets, RCL_RET_INVALID_ARGUMENT);
+  if (0u == num_wait_sets) {
+    RCL_SET_ERROR_MSG("number of wait sets must be greater than zero");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
```
Comment (sloretz) [inline rcl/src/rcl/wait.c]: Could take a short cut and return `rcl_wait()` if `1u == num_wait_sets`

File: rcl/src/rcl/wait.c [hunk 9]
```diff
@@ -614,6 +610,157 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
...
+  // Collate all wait sets into a single wait set
+  for (size_t i = 0u; i < num_wait_sets; ++i) {
+    rcl_wait_set_t * wait_set = &wait_sets[i];
+    for (size_t j = 0u; j < wait_set->size_of_timers; ++j) {
+      const size_t current_index = collated_wait_set.impl->timer_index++;
+      collated_wait_set.timers[current_index] = wait_set->timers[j];
+      rcl_guard_condition_t * guard_condition = rcl_timer_get_guard_condition(wait_set->timers[j]);
+      if (NULL != guard_condition) {
+        // rcl_wait() will take care of moving these backwards and setting guard_condition_count.
+        const size_t offset_rmw_index = wait_set->size_of_guard_conditions + current_index;
```
Comment (sloretz) [inline rcl/src/rcl/wait.c]: `offset_rmw_index` is in the combined wait set, so I think this should be `collated_wait_set->size_of_guard_conditions ...`

File: rcl/test/rcl/test_wait.cpp [hunk 10]
```diff
@@ -435,3 +435,256 @@ TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), guard_condition) {
...
+  const size_t kNumWaitSets = 3u;
+  rcl_wait_set_t wait_sets[kNumWaitSets];
+  rcl_guard_condition_t guard_conds[kNumWaitSets];
+  rcl_timer_t timers[kNumWaitSets];
+  for (size_t i = 0u; i < kNumWaitSets; ++i) {
+    wait_sets[i] = rcl_get_zero_initialized_wait_set();
+    ret = rcl_wait_set_init(&wait_sets[i], 0, 1, 1, 0, 0, rcl_get_default_allocator());
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    // Add a dummy guard condition to avoid an error
```
Comment (sloretz) [inline rcl/test/rcl/test_wait.cpp]: What error is the dummy guard condition avoiding?

File: rcl/src/rcl/wait.c (hunk 9 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: Yeah, good catch!

File: rcl/test/rcl/test_wait.cpp (hunk 10 above)
Comment (jacobperron) [reply rcl/test/rcl/test_wait.cpp]: That is a good question. I copied this over from the existing test involving a timer, without it the test segfaults :fearful:

I dug into it a bit more, and it seems like the rmw guard condition count isn't being incremented in [rcl_wait_set_add_timer()](https://github.com/ros2/rcl/blob/f256da0a86e20c72c6babd394bb34dea697ecbbe/rcl/src/rcl/wait.c#L427). Adding a dummy guard condition appears to be a hack to get the count to increment...

I think it can be fixed easily by ensuring the rmw guard condition count is incremented when adding timers, but without understanding the reason for the current implementation I am hesitant to change it. @wjwwood, is this a bug or intentional?

File: rcl/src/rcl/wait.c (hunk 9 above)
Comment (jacobperron) [reply rcl/src/rcl/wait.c]: Fixed in b58dfa9

File: rcl/test/rcl/test_wait.cpp (hunk 10 above)
Comment (jacobperron) [reply rcl/test/rcl/test_wait.cpp]: Nevermind, I see that the rmw guard condition count is increment for timers in `rcl_wait` as the comment says (my bad):

https://github.com/ros2/rcl/blob/f256da0a86e20c72c6babd394bb34dea697ecbbe/rcl/src/rcl/wait.c#L502

Setting the number of guard conditions to zero resolves the issue. I've also updated  the tests for `rcl_wait()` that originally confused me. See e40a26b

**Comment by jacobperron** [comment]
Reviews addressed.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5567/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2212/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4587/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5457/)

**Comment by hidmic** [comment]
@jacobperron are we pushing this one as-is or are you considering a re-implementation in terms of the feature introduced by #335 ?

**Comment by jacobperron** [comment]
>are we pushing this one as-is or are you considering a re-implementation in terms of the feature introduced by #335 ?

@hidmic I think this feature is not necessary anymore in favor of #335. If you agree, I will close this PR.

**Comment by hidmic** [comment]
Makes total sense to me, let's close it.
